sync.py

```diff
@@ -71,22 +82,64 @@ def all_messages(credentials, full_sync=False) -> int:
-            try:
-                raw_msg = (
-                    service.users().messages().get(userId="me", id=m["id"]).execute()
-                )
-                msg = message.Message.from_raw(raw_msg, labels)
-                db.create_message(msg)
-
-            except IntegrityError as e:
-                print(f"Could not process message {m['id']}: {str(e)}")
-            except TimeoutError as e:
-                print(f"Could not get message from Gmail {m['id']}: {str(e)}")
-
-            print(f"Synced message {msg.id} from {msg.timestamp} (Count: {i})")
+        message_ids = [m["id"] for m in messages]
+
+        # Process in batches (parallel fetch)
+        for batch_start in range(0, len(message_ids), BATCH_SIZE):
+            batch_ids = message_ids[batch_start:batch_start+BATCH_SIZE]
+            print(f"\nProcessing batch {batch_start // BATCH_SIZE + 1} (messages {batch_start + 1}-{batch_start + len(batch_ids)})")
+            print(f"Batch message IDs: {batch_ids}")
+
+            def fetch_message(msg_id):
+                for attempt in range(1, MAX_FETCH_RETRIES + 1):
+                    try:
+                        # Rebuild the service for each thread to avoid sharing connections
+                        local_service = build("gmail", "v1", credentials=credentials, cache_discovery=False)
+                        raw_msg = (
+                            local_service.users()
+                            .messages()
+                            .get(userId="me", id=msg_id, format="full")
+                            .execute()
+                        )
+                        return (msg_id, raw_msg, None)
+                    except HttpError as e:
+                        if attempt < MAX_FETCH_RETRIES:
+                            print(f"HTTP error on attempt {attempt} for message {msg_id}: {e}. Retrying...")
+                            time.sleep(RETRY_SLEEP * attempt)
+                        else:
+                            return (msg_id, None, e)
+                    except Exception as e:
+                        if attempt < MAX_FETCH_RETRIES:
+                            print(f"Error on attempt {attempt} for message {msg_id}: {e}. Retrying...")
+                            time.sleep(RETRY_SLEEP * attempt)
+                        else:
+                            return (msg_id, None, e)
+
+            batch_messages = []
+            with concurrent.futures.ThreadPoolExecutor(max_workers=MAX_THREADS) as executor:
+                future_to_id = {executor.submit(fetch_message, msg_id): msg_id for msg_id in batch_ids}
+                for future in concurrent.futures.as_completed(future_to_id):
+                    msg_id = future_to_id[future]
+                    msg_id, raw_msg, error = future.result()
+                    if error:
+                        print(f"Failed to fetch message {msg_id}: {error}")
+                        continue
+                    batch_messages.append(raw_msg)
```

Comment:
Would it make sense to directly parse and store the message? This would reduce resource usage, and we would not lose messages in the event of an exception.